Add option to find datasets from prefix rather than exact path

Kirill888 opened the pull request:
Adding backwards compatible option to `get_datasets_for_location` called
`mode=prefix|exact|None=auto`.

Default mode is `auto` which means prefix search for uris without fragment part, and exact search if fragment part is present. This way old code that did `index. get_datasets_for_location(path.as_uri())` will still do the same thing even if db was updated with fragment information. Only thing is that this will now return results when path is a valid folder also, when in the past it wouldn't.

### Reason for this pull request

Needed for `dea-sync` to find datasets for stacked netcdfs after we update locations to include `#part=x` in the uri. 



 - [x] Closes #466
 - [x] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

This is internal api as far as I can tell so no additions to `whats_new.rst`

Kirill888 committed:
Add option to find datasets from prefix rather than exact path

Adding backwards compatible option to `get_datasets_for_location` called
`mode=prefix|exact`, default behavior remains as before -- exact match, new
behaviour is to return datasets whose path starts with a supplied uri.

See issue #466.

Kirill888 committed:
Testing exception path for get_dataset_for_location

jeremyh commented:
Thanks Kirill.

Stylistically I'd probably lean towards having two separate methods rather than a `mode` switch. It seems a more common pattern among Python libraries to have explicit methods, and is more convenient to use with editor completion (no need to remember the names of enum-like option strings).

(maybe `get_datasets_for_location_prefix` or  `get_datasets_within_location`?)

I suspect prefix mode will get more usage than the old default: Ops actually wanted to use it yesterday: "find all datasets in a folder".

(The Db method itself is low level, so a mode parameter is ok there if they use the same query code)

Kirill888 committed:
Changed default mode to auto for get_datasets_for_location

Here auto is denoted by `None` and means

1. Exact search if uri contains fragment part
2. Prefix search if it doesn't

This way `index.get_datasets_for_location(file.as_uri())` will do expected thing
of returning all datasets within the file regardless of whether db already
migrated to include fragment part or not.

Kirill888 commented:
After discussions with @jeremyh I have updated code and pull request to default to `auto` mode as opposed to `exact`. The `exact` mode while producing the same query as before would behave in unexpected way on migrated stacked files, `auto` mode will produce different query, but will appear identical in behaviour as before, for the exception of now also working on directories and not just files.

jeremyh reviewed (approved):
:+1:
